File: csharp/Foundation.Data.Doublets.Cli.Tests/VersionControlDecoratorTests.cs
using Platform.Data;
using Platform.Data.Doublets;

using DoubletLink = Platform.Data.Doublets.Link<uint>;

namespace Foundation.Data.Doublets.Cli.Tests.Tests
{
  public class VersionControlDecoratorTests
  {
    [Fact]
    public void DefaultBranchExistsOnFirstOpen()
    {
      RunWithVc((vc, _, _) =>
      {
        Assert.Equal(VersionControlDecorator.DefaultBranchName, vc.CurrentBranch);
        var branches = vc.ListBranches();
        Assert.Single(branches);
        Assert.Equal(VersionControlDecorator.DefaultBranchName, branches[0].Name);
      });
    }

    [Fact]
    public void NewTransitionsAreAttributedToCurrentBranch()
    {
      RunWithVc((vc, tx, _) =>
      {
        var a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var head = tx.LastLoggedSequence;
        Assert.True(head >= 2, $"CreateAndUpdate must produce at least two transitions (got {head}).");
        Assert.Equal(head, vc.CurrentSequence);
      });
    }

    [Fact]
    public void CheckoutToZeroRewindsEverything()
    {
      RunWithVc((vc, tx, _) =>
      {
        var a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var b = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        Assert.True(vc.Exists(a));
        Assert.True(vc.Exists(b));

        vc.Checkout(0);

        Assert.False(vc.Exists(a), "All links must be rewound after checkout 0.");
        Assert.False(vc.Exists(b));
        Assert.Equal(0, vc.CurrentSequence);
      });
    }

    [Fact]
    public void CheckoutAndForwardReplayRestoresState()
    {
      RunWithVc((vc, tx, _) =>
      {
        var a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var afterFirst = tx.LastLoggedSequence;
        var b = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var afterSecond = tx.LastLoggedSequence;

        vc.Checkout(afterFirst);
        Assert.True(vc.Exists(a), "First link must remain after partial rewind.");
        Assert.False(vc.Exists(b), "Second link must disappear after partial rewind.");

        vc.Checkout(afterSecond);
        Assert.True(vc.Exists(a));
        Assert.True(vc.Exists(b), "Second link must reappear after forward checkout.");
      });
    }

    [Fact]
    public void BranchForksFromCurrentHead()
    {
      RunWithVc((vc, tx, _) =>
      {
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var headBeforeBranch = vc.CurrentSequence;

        vc.Branch("feature");
        Assert.Contains(vc.ListBranches(), b => b.Name == "feature");
      });
    }

    [Fact]
    public void SwitchBranchAppliesAndRewindsTransitions()
    {
      RunWithVc((vc, tx, _) =>
      {
        var a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var headBeforeBranch = vc.CurrentSequence;

        vc.Branch("feature");
        vc.SwitchBranch("feature");
        Assert.Equal("feature", vc.CurrentBranch);

        var b = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        Assert.True(vc.Exists(b));
        var featureHead = vc.CurrentSequence;

        vc.SwitchBranch(VersionControlDecorator.DefaultBranchName);
        Assert.Equal(VersionControlDecorator.DefaultBranchName, vc.CurrentBranch);
        Assert.True(vc.Exists(a), "Main-branch link must remain after switching back.");
        Assert.False(vc.Exists(b), "Feature-branch link must disappear after switching back to main.");
        Assert.Equal(headBeforeBranch, vc.CurrentSequence);

        vc.SwitchBranch("feature");
        Assert.True(vc.Exists(a));
        Assert.True(vc.Exists(b), "Feature-branch link must reappear after switching back to feature.");
        Assert.Equal(featureHead, vc.CurrentSequence);
      });
    }

    [Fact]
    public void TagPointsToCurrentHead()
    {
      RunWithVc((vc, tx, _) =>
      {
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        vc.Tag("v1");
        Assert.True(vc.TryGetTag("v1", out var seq));
        Assert.Equal(vc.CurrentSequence, seq);
        Assert.Contains("v1", vc.ListTags().Keys);
      });
    }

    [Fact]
    public void BranchFromExplicitSeqUsesGivenPoint()
    {
      RunWithVc((vc, tx, _) =>
      {
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        var firstHead = vc.CurrentSequence;
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);

        vc.Branch("backport", from: firstHead);
        var branchInfo = vc.ListBranches().Single(b => b.Name == "backport");
        Assert.Equal(firstHead, branchInfo.ForkSeq);
      });
    }

    [Fact]
    public void RecoverRebuildsStateFromBranchesStore()
    {
      // Recovery is exercised here by attaching a *second* VC decorator
      // to the same live branches store, which is equivalent in behaviour
      // to reopening the underlying file (the file-mapped store is shared).
      RunWithVc((vc, _, _) =>
      {
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        vc.Tag("checkpoint");
        vc.Branch("feature");

        // Force a fresh decorator over the same in-process VC store.
        var branchesStore = GetBranchesStore(vc);
        var transactions = GetTransactions(vc);
        var reopened = new VersionControlDecorator(transactions, branchesStore);
        Assert.Contains(reopened.ListBranches(), b => b.Name == "feature");
        Assert.True(reopened.TryGetTag("checkpoint", out _));
      });
    }

    private static INamedTypesLinks<uint> GetBranchesStore(VersionControlDecorator vc)
    {
      return (INamedTypesLinks<uint>)typeof(VersionControlDecorator)
        .GetField("_branchesStore", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)!
        .GetValue(vc)!;
    }

    private static TransactionsDecorator GetTransactions(VersionControlDecorator vc)
    {
      return (TransactionsDecorator)typeof(VersionControlDecorator)
        .GetField("_transactions", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)!
        .GetValue(vc)!;
    }

    [Fact]
    public void CheckoutOutOfRangeThrows()
    {
      RunWithVc((vc, tx, _) =>
      {
        vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
        Assert.Throws<InvalidOperationException>(() => vc.Checkout(999));
      });
    }

    [Fact]
    public void DuplicateBranchThrows()
    {
      RunWithVc((vc, _, _) =>
      {
        vc.Branch("feature");
        Assert.Throws<InvalidOperationException>(() => vc.Branch("feature"));
      });
    }

    [Fact]
    public void FullStackAcidRollbackIsAtomicAndIsolated()
    {
      RunWithVc((vc, _, _) =>
      {
        var baseline = Snapshot(vc);
        var initialSequence = vc.CurrentSequence;

        using (var transaction = vc.BeginTransaction())
        {
          var a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
          var b = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
          vc.Update(
            new DoubletLink(a, vc.Constants.Any, vc.Constants.Any),
            new DoubletLink(a, b, b),
            null);

          Assert.True(vc.Exists(a));
          Assert.True(vc.Exists(b));
          Assert.Throws<InvalidOperationException>(() => vc.BeginTransaction());
          Assert.Throws<InvalidOperationException>(() => vc.Branch("blocked"));

          transaction.Rollback();
        }

        Assert.Equal(initialSequence, vc.CurrentSequence);
        Assert.Equal(initialSequence, vc.ListBranches().Single(b => b.Name == VersionControlDecorator.DefaultBranchName).Head);
        Assert.Equal(baseline, Snapshot(vc));
      });
    }

    [Fact]
    public void FullStackAcidCommitIsConsistentAndDurableAcrossReopen()
    {
      var dataFile = Path.GetTempFileName();
      var logFile = Path.GetTempFileName();
      var vcFile = Path.GetTempFileName();
      NamedTypesDecorator<uint>? dataLinks = null;
      NamedTypesDecorator<uint>? logLinks = null;
      NamedTypesDecorator<uint>? vcLinks = null;
      TransactionsDecorator? tx = null;
      NamedTypesDecorator<uint>? reopenedDataLinks = null;
      NamedTypesDecorator<uint>? reopenedLogLinks = null;
      NamedTypesDecorator<uint>? reopenedVcLinks = null;
      TransactionsDecorator? reopenedTx = null;

      try
      {
        uint a;
        uint b;
        long committedSequence;

        dataLinks = new NamedTypesDecorator<uint>(dataFile);
        logLinks = new NamedTypesDecorator<uint>(logFile);
        vcLinks = new NamedTypesDecorator<uint>(vcFile);
        tx = new TransactionsDecorator(dataLinks, logLinks);
        using var vc = new VersionControlDecorator(tx, vcLinks);

        using (var transaction = vc.BeginTransaction())
        {
          a = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
          b = vc.CreateAndUpdate(vc.Constants.Null, vc.Constants.Null);
          vc.Update(
            new DoubletLink(a, vc.Constants.Any, vc.Constants.Any),
            new DoubletLink(a, b, b),
            null);
          transaction.Commit();
        }

        committedSequence = vc.CurrentSequence;
        Assert.True(committedSequence >= 5);
        Assert.Equal(tx.LastLoggedSequence, tx.AppliedSequence);
        Assert.Equal(committedSequence, vc.ListBranches().Single(branch => branch.Name == VersionControlDecorator.DefaultBranchName).Head);

        vc.Tag("acid-commit");
        vc.Branch("audit");
        vc.SwitchBranch("audit");
        vc.Delete(new DoubletLink(b, vc.Constants.Any, vc.Constants.Any), null);
        Assert.False(vc.Exists(b));

        vc.SwitchBranch(VersionControlDecorator.DefaultBranchName);
        Assert.True(vc.Exists(a));
        Assert.True(vc.Exists(b));
        var restored = new DoubletLink(vc.GetLink(a));
        Assert.Equal(b, restored.Source);
        Assert.Equal(b, restored.Target);

        tx.Dispose();
        tx = null;
        dataLinks.Dispose();
        dataLinks = null;
        logLinks.Dispose();
        logLinks = null;
        vcLinks.Dispose();
        vcLinks = null;

        reopenedDataLinks = new NamedTypesDecorator<uint>(dataFile);
        reopenedLogLinks = new NamedTypesDecorator<uint>(logFile);
        reopenedVcLinks = new NamedTypesDecorator<uint>(vcFile);
        reopenedTx = new TransactionsDecorator(reopenedDataLinks, reopenedLogLinks);
        using var reopened = new VersionControlDecorator(reopenedTx, reopenedVcLinks);

        Assert.True(reopened.TryGetTag("acid-commit", out var tagSequence));
        Assert.Equal(committedSequence, tagSequence);
        Assert.Contains(reopened.ListBranches(), branch => branch.Name == "audit");
        Assert.Equal(VersionControlDecorator.DefaultBranchName, reopened.CurrentBranch);
        Assert.True(reopened.Exists(a));
        Assert.True(reopened.Exists(b));
        restored = new DoubletLink(reopened.GetLink(a));
        Assert.Equal(b, restored.Source);
        Assert.Equal(b, restored.Target);
        Assert.Equal(reopenedTx.LastLoggedSequence, reopenedTx.AppliedSequence);
      }
      finally
      {
        tx?.Dispose();
        reopenedTx?.Dispose();
        dataLinks?.Dispose();
        logLinks?.Dispose();
        vcLinks?.Dispose();
        reopenedDataLinks?.Dispose();
        reopenedLogLinks?.Dispose();
        reopenedVcLinks?.Dispose();
        Cleanup(dataFile);
        Cleanup(logFile);
        Cleanup(vcFile);
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(dataFile));
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(logFile));
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(vcFile));
      }
    }

    private static void RunWithVc(Action<VersionControlDecorator, TransactionsDecorator, NamedTypesDecorator<uint>> action)
    {
      var dataFile = Path.GetTempFileName();
      var logFile = Path.GetTempFileName();
      var vcFile = Path.GetTempFileName();
      NamedTypesDecorator<uint>? dataLinks = null;
      NamedTypesDecorator<uint>? logLinks = null;
      NamedTypesDecorator<uint>? vcLinks = null;
      TransactionsDecorator? tx = null;
      try
      {
        dataLinks = new NamedTypesDecorator<uint>(dataFile);
        logLinks = new NamedTypesDecorator<uint>(logFile);
        vcLinks = new NamedTypesDecorator<uint>(vcFile);
        tx = new TransactionsDecorator(dataLinks, logLinks);
        using var vc = new VersionControlDecorator(tx, vcLinks);
        action(vc, tx, dataLinks);
      }
      finally
      {
        tx?.Dispose();
        dataLinks?.Dispose();
        logLinks?.Dispose();
        vcLinks?.Dispose();
        Cleanup(dataFile);
        Cleanup(logFile);
        Cleanup(vcFile);
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(dataFile));
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(logFile));
        Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(vcFile));
      }
    }

    private static void Cleanup(string path)
    {
      if (File.Exists(path)) File.Delete(path);
    }

    private static IReadOnlyList<DoubletLink> Snapshot(ILinks<uint> links)
    {
      var any = links.Constants.Any;
      var query = new DoubletLink(any, any, any);
      return links.All(query)
        .Select(link => new DoubletLink(link))
        .OrderBy(link => link.Index)
        .ThenBy(link => link.Source)
        .ThenBy(link => link.Target)
        .ToArray();
    }
  }
}

File: rust/tests/version_control_decorator_tests.rs
//! Integration tests for the optional version-control decorator.
//!
//! Mirrors the C# `VersionControlDecoratorTests` and exercises R11–R16
//! of issue #94.

use anyhow::Result;
use link_cli::{
    CommitMode, LogRetentionPolicy, NamedTypesDecorator, TransactionsDecorator,
    VersionControlDecorator, DEFAULT_BRANCH_NAME,
};
use tempfile::NamedTempFile;

struct VcGuard {
    _data_file: NamedTempFile,
    _log_file: NamedTempFile,
    _vc_file: NamedTempFile,
}

fn make_vc() -> (VersionControlDecorator, VcGuard) {
    let data_file = NamedTempFile::new().unwrap();
    let log_file = NamedTempFile::new().unwrap();
    let vc_file = NamedTempFile::new().unwrap();
    let data_links = NamedTypesDecorator::new(data_file.path(), false).unwrap();
    let log_links = NamedTypesDecorator::new(log_file.path(), false).unwrap();
    let vc_links = NamedTypesDecorator::new(vc_file.path(), false).unwrap();
    let tx = TransactionsDecorator::new(
        data_links,
        log_links,
        LogRetentionPolicy::default(),
        CommitMode::default(),
        false,
    )
    .unwrap();
    let vc = VersionControlDecorator::new(tx, vc_links, false).unwrap();
    (
        vc,
        VcGuard {
            _data_file: data_file,
            _log_file: log_file,
            _vc_file: vc_file,
        },
    )
}

#[test]
fn default_branch_exists_on_first_open() {
    let (vc, _guard) = make_vc();
    assert_eq!(DEFAULT_BRANCH_NAME, vc.current_branch());
    let branches = vc.list_branches();
    assert_eq!(1, branches.len());
    assert_eq!(DEFAULT_BRANCH_NAME, branches[0].name);
}

#[test]
fn new_transitions_are_attributed_to_current_branch() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    let _id = vc.create_and_update(0, 0)?;
    let head = vc.transactions().last_logged_sequence();
    assert!(
        head >= 2,
        "create_and_update must produce at least two transitions (got {head})."
    );
    assert_eq!(head, vc.current_sequence());
    Ok(())
}

#[test]
fn checkout_to_zero_rewinds_everything() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    let a = vc.create_and_update(0, 0)?;
    let b = vc.create_and_update(0, 0)?;
    assert!(vc.exists(a));
    assert!(vc.exists(b));

    vc.checkout(0)?;

    assert!(!vc.exists(a), "all links must be rewound after checkout 0");
    assert!(!vc.exists(b));
    assert_eq!(0, vc.current_sequence());
    Ok(())
}

#[test]
fn checkout_and_forward_replay_restores_state() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    let a = vc.create_and_update(0, 0)?;
    let after_first = vc.transactions().last_logged_sequence();
    let b = vc.create_and_update(0, 0)?;
    let after_second = vc.transactions().last_logged_sequence();

    vc.checkout(after_first)?;
    assert!(vc.exists(a), "first link must remain after partial rewind");
    assert!(!vc.exists(b), "second link must disappear after partial rewind");

    vc.checkout(after_second)?;
    assert!(vc.exists(a));
    assert!(vc.exists(b), "second link must reappear after forward checkout");
    Ok(())
}

#[test]
fn branch_forks_from_current_head() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    vc.create_and_update(0, 0)?;
    vc.branch("feature", None)?;
    assert!(vc.list_branches().iter().any(|b| b.name == "feature"));
    Ok(())
}

#[test]
fn switch_branch_applies_and_rewinds_transitions() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    let a = vc.create_and_update(0, 0)?;
    let head_before_branch = vc.current_sequence();

    vc.branch("feature", None)?;
    vc.switch_branch("feature")?;
    assert_eq!("feature", vc.current_branch());

    let b = vc.create_and_update(0, 0)?;
    assert!(vc.exists(b));
    let feature_head = vc.current_sequence();

    vc.switch_branch(DEFAULT_BRANCH_NAME)?;
    assert_eq!(DEFAULT_BRANCH_NAME, vc.current_branch());
    assert!(vc.exists(a), "main-branch link must remain after switching back");
    assert!(!vc.exists(b), "feature-branch link must disappear after switching back to main");
    assert_eq!(head_before_branch, vc.current_sequence());

    vc.switch_branch("feature")?;
    assert!(vc.exists(a));
    assert!(vc.exists(b), "feature-branch link must reappear");
    assert_eq!(feature_head, vc.current_sequence());
    Ok(())
}

#[test]
fn tag_points_to_current_head() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    vc.create_and_update(0, 0)?;
    vc.tag("v1", None)?;
    let seq = vc.try_get_tag("v1").expect("tag must be retrievable");
    assert_eq!(vc.current_sequence(), seq);
    assert!(vc.list_tags().contains_key("v1"));
    Ok(())
}

#[test]
fn branch_from_explicit_seq_uses_given_point() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    vc.create_and_update(0, 0)?;
    let first_head = vc.current_sequence();
    vc.create_and_update(0, 0)?;

    vc.branch("backport", Some(first_head))?;
    let branches = vc.list_branches();
    let branch = branches
        .iter()
        .find(|b| b.name == "backport")
        .expect("backport branch must exist");
    assert_eq!(first_head, branch.fork_seq);
    Ok(())
}

#[test]
fn recover_rebuilds_state_from_branches_store() -> Result<()> {
    let data_file = NamedTempFile::new()?;
    let log_file = NamedTempFile::new()?;
    let vc_file = NamedTempFile::new()?;
    let data_path = data_file.path().to_path_buf();
    let log_path = log_file.path().to_path_buf();
    let vc_path = vc_file.path().to_path_buf();

    {
        let data_links = NamedTypesDecorator::new(&data_path, false)?;
        let log_links = NamedTypesDecorator::new(&log_path, false)?;
        let vc_links = NamedTypesDecorator::new(&vc_path, false)?;
        let tx = TransactionsDecorator::new(
            data_links,
            log_links,
            LogRetentionPolicy::default(),
            CommitMode::default(),
            false,
        )?;
        let mut vc = VersionControlDecorator::new(tx, vc_links, false)?;
        vc.create_and_update(0, 0)?;
        vc.tag("checkpoint", None)?;
        vc.branch("feature", None)?;
        vc.save()?;
    }

    let data_links = NamedTypesDecorator::new(&data_path, false)?;
    let log_links = NamedTypesDecorator::new(&log_path, false)?;
    let vc_links = NamedTypesDecorator::new(&vc_path, false)?;
    let tx = TransactionsDecorator::new(
        data_links,
        log_links,
        LogRetentionPolicy::default(),
        CommitMode::default(),
        false,
    )?;
    let reopened = VersionControlDecorator::new(tx, vc_links, false)?;
    assert!(reopened
        .list_branches()
        .iter()
        .any(|b| b.name == "feature"));
    assert!(reopened.try_get_tag("checkpoint").is_some());
    Ok(())
}

#[test]
fn checkout_out_of_range_throws() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    vc.create_and_update(0, 0)?;
    assert!(vc.checkout(999).is_err());
    Ok(())
}

#[test]
fn duplicate_branch_throws() -> Result<()> {
    let (mut vc, _guard) = make_vc();
    vc.branch("feature", None)?;
    assert!(vc.branch("feature", None).is_err());
    Ok(())
}
